lib/microcbor-derive/src/lib.rs

```diff
@@ -0,0 +1,1030 @@
+///   representation of the variant, instead.
+///
+///   This attribute may only be placed on unit variants, unless the enum type
+///   uses the [variant-ID representation](#enum-variant-id-encoding).
+///
+/// [serde-tagged]: https://serde.rs/enum-representations.html#internally-tagged
+/// [cbor-tags]: https://www.rfc-editor.org/rfc/rfc8949.html#name-tagging-of-items
+#[proc_macro_derive(Encode, attributes(cbor))]
+pub fn derive_encode(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    match gen_encode_impl(input) {
+        Ok(tokens) => tokens,
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+/// Derives an implementation of the [`EncodeFields`] trait for the annotated
+/// `struct` or `enum` type.
+///
+/// Deriving `EncodeFields` allows the implementing type to be annotated with
+/// `#[cbor(flatten)]` when nested within another type that derives `Encode` or
+/// `EncodeFields`.
+///
+/// Types that derive `EncodeFields` may only have named fields. If the deriving
+/// type is an `enum`, all variants must have named fields; attempting to derive
+/// `EncodeFields` on an enum that has both named (struct-like) variants and
+/// unnamed (tuple-like) variants will result in a compilation error.
+///
+/// The same type may derive both `Encode` and `EncodeFields` to be able to be
+/// encoded both as its own map and flattened into existing maps.
+///
+/// # Helper Attributes
+///
+/// All [the attributes](macro@Encode#helper-attributes) recognized by
+/// `#[derive(Encode)]` may also be placed on the fields of a type that derives
+/// `EncodeFields`.
+#[proc_macro_derive(EncodeFields, attributes(cbor))]
+pub fn derive_encode_fields(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    match gen_encode_fields_impl(input) {
+        Ok(tokens) => tokens,
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+fn gen_encode_impl(input: DeriveInput) -> Result<TokenStream, syn::Error> {
+    match &input.data {
+        syn::Data::Enum(data) => gen_enum_encode_impl(
+            input.attrs,
+            input.vis,
+            input.ident,
+            input.generics,
+            data.clone(),
+        )
+        .map(|tokens| tokens.to_token_stream().into()),
+        syn::Data::Struct(data) => gen_encode_struct_impl(
+            input.attrs,
+            input.vis,
+            input.ident,
+            input.generics,
+            data.clone(),
+        )
+        .map(|tokens| tokens.to_token_stream().into()),
+        _ => Err(syn::Error::new_spanned(
+            input,
+            "`StaticCborLen` can only be derived for `struct` and `enum` \
+             types",
+        )),
+    }
+}
+
+const HELPER_ATTR: &str = "cbor";
+const RENAME_ATTR: &str = "rename";
+const VARIANT_ID_ATTR: &str = "variant_id";
+
+fn gen_enum_encode_impl(
+    attrs: Vec<Attribute>,
+    _vis: Visibility,
```

Comment:
Is there a reason to keep the dead arguments in this and other `gen_` functions?